sensing/src/main/java/com/google/android/sensing/impl/SensingEngineImpl.kt

```diff
@@ -212,6 +213,56 @@ internal class SensingEngineImpl(
     TODO("Not yet implemented")
   }
 
+  override suspend fun resetFailedUploadRequests() {
+    val failedUploadRequestList: List<UploadRequest>  = database.listUploadRequests(status = RequestStatus.FAILED)
+    // Extract the resourceInfoIds from pending requests for quick lookup
+    val pendingResourceIds = database.listUploadRequests(status = RequestStatus.PENDING).map { it.resourceInfoId }.toSet()
+    // Extract the resourceInfoIds from uploaded requests for quick lookup
+    val uploadedResourceIds = database.listUploadRequests(status = RequestStatus.UPLOADED).map { it.resourceInfoId }.toSet()
+    // Filter failed requests whose resourceInfoId is not in pending and completed requests
+    val failedNotInPendingOrCompletedList = failedUploadRequestList
+      .filter { it.resourceInfoId !in pendingResourceIds && it.resourceInfoId !in uploadedResourceIds }
+
+    if (failedNotInPendingOrCompletedList.isNotEmpty()) {
+      failedNotInPendingOrCompletedList.forEach{ failedNotInPendingOrCompleted ->
+        val uploadRequest =
+          UploadRequest(
+            requestUuid = UUID.randomUUID(),
+            resourceInfoId = failedNotInPendingOrCompleted.resourceInfoId,
+            zipFile = failedNotInPendingOrCompleted.zipFile,
+            fileSize = failedNotInPendingOrCompleted.fileSize,
+            bucketName = failedNotInPendingOrCompleted.bucketName,
+            uploadRelativeURL = failedNotInPendingOrCompleted.uploadRelativeURL,
+            isMultiPart = failedNotInPendingOrCompleted.isMultiPart
+          )
+        with(File(uploadRequest.zipFile)) {
+          if (!this.exists()) {
+            val folderPath = this.absolutePath.substring(0, this.absolutePath.length - 4)
+            if (!File(folderPath).exists()) throw FileNotFoundException()
+            // zip the folder using same logic in SensingEngineImpl.
+            /** Zipping logic from: https://stackoverflow.com/a/63828765 */
+            val resourceFolder = File(folderPath)
+            val zipOutputStream =
+              ZipOutputStream(BufferedOutputStream(FileOutputStream(this.absolutePath)))
+            zipOutputStream.use { zos ->
+              resourceFolder.walkTopDown().forEach { file ->
+                val zipFileName =
+                  file.absolutePath.removePrefix(resourceFolder.absolutePath).removePrefix("/")
+                val entry = ZipEntry("$zipFileName${(if (file.isDirectory) "/" else "")}")
+                zos.putNextEntry(entry)
+                if (file.isFile) {
+                  file.inputStream().use { fis -> fis.copyTo(zos) }
+                }
+              }
+            }
+          }
+        }
```

Comment:
can you make a utility function of zipping a folder given a folder path ?
And use that function here as well as in the other function - `onCaptureCompleteCallback`